src/napi.h

```diff
@@ -0,0 +1,558 @@
+﻿#ifndef SRC_NODE_API_HELPERS_H_
+#define SRC_NAPI_H_
+
+////////////////////////////////////////////////////////////////////////////////
+// NAPI C++ Wrapper Classes
+//
+// These classes wrap the "NAPI" ABI-stable C APIs for Node.js, providing a
+// C++ object model and C++ exception-handling semantics with low overhead.
+// The wrappers are all header-only so that they do not affect the ABI.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "node_jsvmapi.h"
+#include "node_asyncapi.h"
+#include <string>
+#include <vector>
+
+namespace Napi {
+
+  class Value;
+  class Boolean;
+  class Number;
+  class String;
+  class Object;
+  class Array;
+  class Function;
+  class Buffer;
+  class Error;
+  class Persistent;
+  class PropertyDescriptor;
+  class CallbackInfo;
+
+  // Functions exposed to JavaScript must conform to one these callback signatures.
+  // (See ObjectWrap<T> for callbacks used when wrapping entire classes.)
+  typedef void (*VoidFunctionCallback)(const CallbackInfo& info);
+  typedef Value (*FunctionCallback)(const CallbackInfo& info);
+
+  /*
+   * Environment for NAPI operations. (In V8 this corresponds to an Isolate.)
+   */
+  class Env {
+  public:
+    explicit Env(napi_env env);
+
+    operator napi_env() const;
+
+    static Env Current();
+
+    Object Global() const;
+    Value Undefined() const;
+    Value Null() const;
+
+    Error NewError();
+    Error NewError(const char* message);
+    Error NewTypeError(const char* message);
+    Buffer NewBuffer(char* data, size_t size);
+    Buffer CopyBuffer(const char* data, size_t size);
+    Number NewNumber(double val);
+    Boolean NewBoolean(bool val);
+    String NewStringUtf8(const char* val, int length = -1);
+    String NewStringUtf16(const char16_t* val, int length = -1);
+    Array NewArray();
+    Array NewArray(int length);
+    Object NewObject();
```

Comment:
It may be helpful to review my typed-array PRs first, since I'm also going to rebase this PR on top of those changes so I can include typed-array support in the wrapper classes.